lms/services/digest.py

```diff
@@ -0,0 +1,264 @@
+
+        return bool(
+            db.scalar(
+                select(func.count())
+                .select_from(AssignmentMembership)
+                .join(Assignment)
+                .join(AssignmentGrouping)
+                .join(LTIRole)
+                .filter(AssignmentGrouping.grouping_id.in_(course_ids))
+                .filter(AssignmentMembership.user_id.in_(user.user_ids))
+                .filter(LTIRole.type == "instructor", LTIRole.scope == "course")
+            )
+        )
+
+    @staticmethod
+    def get_courses(db, authority_provided_id):
+        """Return a list of all Course's with the given authority_provided_id."""
+        return db.scalars(
+            select(Course)
+            .filter_by(authority_provided_id=authority_provided_id)
+            .order_by(Course.updated.desc())
+        ).all()
+
+    @classmethod
+    def make(cls, db, authority_provided_id):
+        """Return a UnifiedCourse for the given authority_provided_id."""
+        first_grouping = db.scalars(
+            select(Grouping)
+            .filter_by(authority_provided_id=authority_provided_id)
+            .order_by(Grouping.updated.desc())
+        ).first()
+
+        if not first_grouping:
+            raise UnknownAuthorityProvidedID()
+
+        if first_grouping.type == Grouping.Type.COURSE:
+            authority_provided_id = first_grouping.authority_provided_id
+        else:
+            assert first_grouping.parent.type == Grouping.Type.COURSE
+            authority_provided_id = first_grouping.parent.authority_provided_id
+
+        return cls(cls.get_courses(db, authority_provided_id))
+
+
+class DigestHelper:
+    """Helper methods for DigestService."""
+
+    def __init__(self, db):
+        self.db = db
+
+    def unified_courses(self, annotations) -> List[UnifiedCourse]:
+        """
+        Return the given list of annotation dicts organized by course.
+
+        >>> helper.annotations_by_course([{annotation}, ...])
+        [
+            UnifiedCourse(authority_provided_id="id1", annotations=[{annotation}, ...]),
+            UnifiedCourse(authority_provided_id="id2", annotations=[{annotation}, ...]),
+            ...
+        }
+        """
+        unified_courses = {}
+
+        for annotation in annotations:
+            authority_provided_id = annotation["group"]["authority_provided_id"]
+
+            try:
+                unified_course = UnifiedCourse.make(self.db, authority_provided_id)
+            except UnknownAuthorityProvidedID:
+                continue
+
+            unified_course = unified_courses.setdefault(
+                unified_course.authority_provided_id, unified_course
+            )
+            unified_course.annotations.append(annotation)
+
+        return list(unified_courses.values())
+
+    def instructor_digest(
+        self, instructor: UnifiedUser, unified_courses: List[UnifiedCourse]
+    ):
+        """
+        Return a digest for the given instructor and unified_courses.
```

Comment:
Maybe word this so it refers to the "_variables_" needed for the current email digest template.

Reply:
Changed it to "Return a dict of digest variables..."